File: Go/include/game.go
package include

type Game struct {
	AppleCount int
	Snake      *Snake
	Apple      *Apple
	GUI        *GUI
}

func NewGame(startX, startY int) *Game {
	snake := NewSnake(startX, startY, NewNode(startX, startY+1, nil))
	apple := NewApple(snake)
	gui := NewGUI()
	return &Game{0, snake, apple, gui}
}

func Start(game *Game) {
	HideCursor()

	snake := game.Snake
	apple := game.Apple
	gui := game.GUI

	dx, dy := 1, 0
	running := true

	for {
		input := uint8(GetKey())

		switch input {
		case 'w', 'W', 72:
			if dy != 1 {
				dx, dy = 0, -1
			}
		case 's', 'S', 80:
			if dy != -1 {
				dx, dy = 0, 1
			}
		case 'a', 'A', 75:
			if dx != 1 {
				dx, dy = -1, 0
			}
		case 'd', 'D', 77:
			if dx != -1 {
				dx, dy = 1, 0
			}
		case 'q', 'Q', 27:
			running = false
		}

		if !running {
			break
		}

		ax := snake.Head.X + dx
		ay := snake.Head.Y + dy

		grow := (ax == apple.X && ay == apple.Y)

		if grow {
			Eat(apple, snake)
			game.AppleCount++
		}

		if !MoveSnake(snake, ax, ay, grow) {
			break
		}

		GenGUI(gui, snake, apple, game.AppleCount)

		SleepMS(300)
	}
	GameOver()
	ShowCursor()
	WaitForExit()
}

File: CSharp/include/Game.cs
namespace Ouroboros.Include;

public class Game
{
    private int Score;

    private readonly Snake Snake;

    private readonly Apple Apple;

    private readonly GUI GUI;

    public Game(int start_x, int start_y)
    {
        Score = 0;
        Snake = new(start_x, start_y, new(start_x, start_y + 1));
        Apple = new(Snake);
        GUI = new();
    }

    public void Start()
    {
        Terminal.HideCursor();

        int dx = 1;
        int dy = 0;
        int ax, ay;

        for (;;)
        {
            var input = Input.GetKeyPress() ?? ConsoleKey.NoName;

            (dx, dy) = input switch
            {
                ConsoleKey.W or ConsoleKey.UpArrow when dy != 1 => (0, -1),

                ConsoleKey.S or ConsoleKey.DownArrow when dy != -1 => (0, 1),

                ConsoleKey.A or ConsoleKey.LeftArrow when dx != 1 => (-1, 0),

                ConsoleKey.D or ConsoleKey.RightArrow when dx != -1 => (1, 0),

                _ => (dx, dy)
            };

            if (input is ConsoleKey.Q or ConsoleKey.Escape)
            {
                break;
            }

            (ax, ay) = (Snake.Head.X + dx, Snake.Head.Y + dy);

            var grow = (ax == Apple.X && ay == Apple.Y);

            if (grow)
            {
                Apple.Eat(Snake);
                ++Score;
            }

            if (!Snake.Move(ax, ay, grow))
            {
                break;
            }

            GUI.GenGUI(Snake, Apple, Score);
            Time.SleepMs(300);
        }
        GUI.GameOver();
        Terminal.ShowCursor();
        Input.WaitForExit();
    }
}